### PR TITLE
Develop upstream sync 190328

### PR DESCRIPTION
**Pull request by deven-amd**

### CODE REVIEW

**Comment by deven-amd** [comment]
@parallelo 

something seems to be broken in the cuda setup...this is the error we get

```
The following packages have unmet dependencies:
 libnvinfer-dev : Depends: libnvinfer5 (= 5.0.2-1+cuda10.0) but 5.1.2-1+cuda10.1 is to be installed
[91mE: Unable to correct problems, you have held broken packages.
[0mThe command '/bin/sh -c apt-get update         && apt-get install nvinfer-runtime-trt-repo-ubuntu1604-5.0.2-ga-cuda10.0         && apt-get update         && apt-get install -y --no-install-recommends libnvinfer-dev=5.0.2-1+cuda10.0' returned a non-zero code: 100
ERROR: docker build failed. Dockerfile is at /home/ubuntu/workspace/tensorflow-upstream-unit-tests-nvidia/tensorflow/tools/ci_build/Dockerfile.gpu
Build step 'Execute shell' marked build as failure
Setting status of 0d48cb4e56f1a07ce42c77f7ae7f3d552d8c8bc1 to FAILURE with url http://ml-ci.amd.com:21096/job/tensorflow-upstream-unit-tests-nvidia/541/ and message: 'Build finished. '
Using context: cuda
Finished: FAILURE
````

this PR is for the upstream merge, guessing something changed in the upstream w.r.t. the CUDA setup.

Note that we had a successfull CUDA run just an hour ago on a separate PR (which does not include the contents merged from upstream), and hence the suspicion that an upstream change maybe triggering this error.

thanks

**Comment by parallelo** [comment]
I'll take a look to see what changed

File: tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc [hunk 1]
```diff
@@ -130,4 +138,4 @@ template struct UpdateVariableAndFill_Philox<
 
 }  // end namespace tensorflow
 
-#endif  // GOOGLE_CUDA
+#endif  // GOOGLE_CUDA || TENSORFLOR_USE_ROCM
```
Comment (whchung) [inline tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc]: nit picking: typo

File: tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc [hunk 2]
```diff
@@ -75,10 +79,14 @@ void UpdateVariableAndFill_Philox<GPUDevice, Distribution>::operator()(
                                              FillKernel<Distribution>, 0, 0);
 
   int zero = 0;
+#if GOOGLE_CUDA  
   cudaMemcpyToSymbol(thread_counter, &zero, sizeof(int));
+#else // TENSORFLOW_USE_ROCM#
+  hipMemcpyToSymbol(HIP_SYMBOL(thread_counter), &zero, sizeof(int));
```
Comment (whchung) [inline tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc]: hopefully this op would stay as-is for some time before moving to become fully dynamically loaded...

File: tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/core/kernels/stateful_random_ops_gpu.cu.cc]: will fix this, the next time I need to commit something (which hopefully wont be for this PR!)

**Comment by deven-amd** [comment]
So there are three failures in `rocm-xla`
```
//tensorflow/compiler/tests:xla_device_gpu_test                          FAILED in 3 out of 3 in 10.7s                                                                                                                                                                           
//tensorflow/compiler/tests:xla_device_gpu_test_gpu                      FAILED in 3 out of 3 in 11.1s                                                                                                                                                                           
//tensorflow/compiler/tf2xla:fused_batchnorm_reserve_space_test_gpu      FAILED in 3 out of 3 in 6.1s                                                                                                                                                                            
```

All 3 failures, pass consitently for me locally :(

Hoping they are flaky and pass on the next run

----------------------------------------------

On the `cuda` path, I will push a commit that fixes the compiler errors in 
`tensorflow/core/kernels/tridiagonal_solve_op_gpu.cu.cc`, 
but that is only half the problem. 

My local compile still fails because it seems I now need to have correct version of "TensorRT" installed on my local machine, which I do not believe I have. The error I run into is the following:
```
tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc: In function 'void tensorflow::tensorrt::convert::InitializeTrtPlugins()':
tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1032:61: error: 'initLibNvInferPlugins' was not declared in this scope
   plugin_initialized = initLibNvInferPlugins(&trt_logger, "");
                                                             ^
tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1039:3: error: 'IPluginCreator' is not a member of 'nvinfer1'
   nvinfer1::IPluginCreator* const* trt_plugin_creator_list =
   ^
tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1039:29: error: expected primary-expression before 'const'
   nvinfer1::IPluginCreator* const* trt_plugin_creator_list =
                             ^
tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1041:8: error: 'trt_plugin_creator_list' was not declared in this scope
   if (!trt_plugin_creator_list) {
        ^
Target //tensorflow/tools/pip_package:build_pip_package failed to build
INFO: Elapsed time: 173.707s, Critical Path: 158.49s
INFO: 751 processes: 751 local.
FAILED: Build did NOT complete successfully
```

So will push a commit for the part of the CUDA fix I do have, in the hopes that the CI node has all the cuda pre-reqs installed correctly and won't run into the compile error I see locally.

**Comment by deven-amd** [comment]
hmmm...something funky is going on with `rocm-xla`

The three tests that failed in the previous run, and this time around there was only one failure...a compile failure

```
Execution platform: @bazel_tools//platforms:host_platform                                                                                                                                                                                                                        
In file included from tensorflow/compiler/xrt/client/xrt_grpc_eager_client.cc:16:0:                                                                                                                                                                                              
./tensorflow/compiler/xrt/client/xrt_grpc_eager_client.h:26:65: fatal error: net/grpc/public/include/grpcpp/generic/generic_stub.h: No such file or directory                                                                                                                    
compilation terminated.                                                                                                                                                                                                                                                          
```

Now that I look back at the previous log, same build error was present in that one too...ugh. 
seems to be a versioning issue again, as the missing header file seems to be from an external package (`grpc`)

**Comment by parallelo** [comment]
FYI - I just fixed an early abort issue with `rocm-v2` (separate issue from your most recent comment)

**Comment by deven-amd** [comment]
will go ahead with the merge since all three of `rocm` `rocm-xla` and `cuda` paths pass.

`rocm-v2` has 17 failures + 1 build error, but we just added that path and do not know the status for that path. I will file a new PR next week to iron out that path, once I have had a chance to look at failures and either fix them or put them on the whitelist.

**Pull request merged**